rfd/0227-ec2-discovery-by-aws-org.md

```diff
@@ -0,0 +1,442 @@
+This limitation comes from the [`organization.ListAccounts` API](https://docs.aws.amazon.com/organizations/latest/APIReference/API_ListAccounts.html):
+> This operation can be called only from the organization's management account or by a member account that is a delegated administrator.
+
+EC2 instances that exist in the Accounts under the organization specified will join the cluster, and users can access them from Teleport.
+
+### Access EC2 instances of a given AWS Organization
+
+Two flows must be changed in order to achieve the RFD goal:
+- EC2 server discovery must be able to enumerate all the account ids and install teleport into them
+- IAM Join flow must accept instances from any account of a given organization
+
+#### EC2 Server Discovery: install teleport in all the EC2 instances of a given organization
+
+When discovering EC2 instances to install teleport into, the Discovery Service will call [`organizations.ListAccounts`](https://docs.aws.amazon.com/organizations/latest/APIReference/API_ListAccounts.html).
+This API returns all the Accounts under the current organization.
+
+Using the Assume Role feature, which already exists, the service will assume a role in each AWS Account and run the same process it has today.
+
+```go
+func (f *ec2InstanceFetcher) GetInstances(ctx context.Context, rotation bool) ([]Instances, error) {
+  // ...
+  var assumeRoles []*types.AssumeRole
+  accountList, err := organizationsClient.ListAccounts(ctx, &organizations.ListAccountsInput{})
+  for _, account := range accountList.Accounts {
+    assumeRoleARNParts.AccountID = aws.ToString(account.Id)
+
+    assumeRoles = append(assumeRoles, &types.AssumeRole{
+      RoleARN:    assumeRoleARNParts.String(),
+      ExternalID: f.AssumeRole.ExternalID,
+    })
+  }
+
+  // ...
+  for _, assumeRole := range assumeRoles {
+    ec2Client, err := f.EC2Getter(ctx, f.Region, assumeRole, f.AWSClientsOpts...)
+    if err != nil {
+      return nil, trace.Wrap(err)
+    }
+
+    // use ec2Client to call ec2.DescribeInstances and submit instances
+    // afterwards, ssm.SendCommand will be used to install teleport into them
+  }
+}
+```
+
+### Teleport IAM Join based on AWS Organization
+
+In order for an instance to join Teleport using the IAM Join method, the instance signs an `sts.GetCallerIdentity` request and sends it to the Auth service.
+The Auth executes the request on behalf of the instance.
+The response contains the Account ID, and that's how Teleport ensures only instances whose account id is the same as the IAM Join Token, are accepted.
+
+This RFD aims to add a small change to this flow: allow validation to be based on the AWS Organization.
+
+When an EC2 instance tries to join the cluster using the IAM method, it will:
+- create a signed request for `sts.GetCallerIdentity`, which already happens today
+- create a signed request for `organizations.DescribeOrganization`
```

Comment:
I think we wanted to remove this in favor of having the Discovery service query organizations to validate the instance instead using either its ambient or integration credentials?